pkg/controller/oneagent/oneagent_update.go

```diff
@@ -0,0 +1,211 @@
+		// Check if pods have latest agent version
+		outdatedPods, err := r.findOutdatedPodsImmutableImage(r.logger, instance, isLatest)
+		if len(outdatedPods) > 0 {
+			updateCR = true
+		}
+		if err != nil {
+			return updateCR, err
+		}
+
+		err = r.deletePods(r.logger, outdatedPods, buildLabels(instance.GetName()), waitSecs)
+		if err != nil {
+			r.logger.Error(err, err.Error())
+			return updateCR, err
+		}
+		instance.GetOneAgentStatus().UpdatedTimestamp = metav1.Now()
+		err = r.client.Status().Update(context.TODO(), instance)
+		if err != nil {
+			r.logger.Info("failed to updated instance status", "message", err.Error())
+		}
+	} else if instance.GetOneAgentSpec().DisableAgentUpdate {
+		r.logger.Info("Skipping updating pods because of configuration", "disableOneAgentUpdate", true)
+	}
+	return updateCR, nil
+}
+
+// findOutdatedPodsInstaller determines if a pod needs to be restarted in order to get the desired agent version
+// Returns an array of pods and an array of OneAgentInstance objects for status update
+func findOutdatedPodsInstaller(pods []corev1.Pod, dtc dtclient.Client, instance dynatracev1alpha1.BaseOneAgentDaemonSet) ([]corev1.Pod, map[string]dynatracev1alpha1.OneAgentInstance, error) {
+	var doomedPods []corev1.Pod
+	instances := make(map[string]dynatracev1alpha1.OneAgentInstance)
+
+	for _, pod := range pods {
+		item := dynatracev1alpha1.OneAgentInstance{
+			PodName:   pod.Name,
+			IPAddress: pod.Status.HostIP,
+		}
+		ver, err := dtc.GetAgentVersionForIP(pod.Status.HostIP)
+		if err != nil {
+			var serr dtclient.ServerError
+			if ok := errors.As(err, &serr); ok && serr.Code == http.StatusTooManyRequests {
+				return nil, nil, err
+			}
+			// use last know version if available
+			if i, ok := instance.GetOneAgentStatus().Instances[pod.Spec.NodeName]; ok {
+				item.Version = i.Version
+			}
+		} else {
+			item.Version = ver
+			if ver != instance.GetOneAgentStatus().Version {
+				doomedPods = append(doomedPods, pod)
+			}
+		}
+		instances[pod.Spec.NodeName] = item
+	}
+
+	return doomedPods, instances, nil
+}
+
+func (r *ReconcileOneAgent) findOutdatedPodsImmutableImage(logger logr.Logger, instance dynatracev1alpha1.BaseOneAgentDaemonSet, isLatestFn func(logr.Logger, *corev1.ContainerStatus, *corev1.Secret) (bool, error)) ([]corev1.Pod, error) {
+	pods, err := r.findPods(instance)
+	if err != nil {
+		logger.Error(err, "failed to list pods")
+		return nil, err
+	}
+
+	var outdatedPods []corev1.Pod
+	for _, pod := range pods {
+		for _, status := range pod.Status.ContainerStatuses {
+			if status.Image == "" || status.ImageID == "" {
+				// If image is not yet pulled skip check
+				continue
+			}
+			logger.Info("pods container status", "pod", pod.Name, "container", status.Name, "image id", status.ImageID)
+
+			imagePullSecret := &corev1.Secret{}
+			err := r.client.Get(context.TODO(), client.ObjectKey{Namespace: pod.Namespace, Name: instance.GetName() + "-pull-secret"}, imagePullSecret)
```

Comment:
You could technically generate the DockerConfig object from the tokens, rather than creating the Secret on Kubernetes and then keep querying it.

That being said, if there is a custom image, then we may want to check for a custom pull secret.